File: ruby/lib/swiftauth.rb
require "json"
require "net/http"
require "uri"
require "openssl"
require "socket"
require "zlib"
require "digest"

module SwiftAuth
  class Error < StandardError
    attr_reader :code

    def initialize(code, message)
      @code = code
      super("[#{code}] #{message}")
    end
  end

  AppInfo = Struct.new(:name, :version, :anti_debug, :anti_vm, :lock_hwid, :lock_ip, :lock_pc_name, keyword_init: true)
  UserData = Struct.new(:key, :username, :email, :level, :expires_at, :metadata, keyword_init: true)
  Variable = Struct.new(:key, :value, :type, keyword_init: true)
  UserVariable = Struct.new(:key, :value, keyword_init: true)

  class Client
    attr_reader :app, :user, :session_token

    def initialize(base_url:, app_secret:, app_version: "1.0.0", hwid: nil)
      @base_url = base_url.chomp("/")
      @secret = app_secret
      @version = app_version
      @hwid = hwid || "#{Socket.gethostname}-#{ENV["USER"] || ENV["USERNAME"] || "unknown"}"
      @session_token = nil
      @nonce = nil
      @app = nil
      @user = nil
      @ws_callbacks = Hash.new { |h, k| h[k] = [] }
    end

    def initialized?
      !@session_token.nil?
    end

    # ── Initialization ──────────────────────────────────────────────

    def init
      data = post("/api/client/init", {
        secret: @secret,
        version: @version,
        hwid: @hwid,
      })
      @session_token = data["sessionToken"]
      @app = AppInfo.new(
        name: data["appName"] || "",
        version: data["appVersion"] || "",
        anti_debug: data["antiDebug"] || false,
        anti_vm: data["antiVM"] || false,
        lock_hwid: data["lockHwid"] || false,
        lock_ip: data["lockIp"] || false,
        lock_pc_name: data["lockPcName"] || false,
      )
      data
    end

    # ── Authentication ──────────────────────────────────────────────

    def login(username, password, license_key: "", pc_name: "")
      require_init!
      fetch_nonce
      data = post_with_nonce("/api/client/login", {
        sessionToken: @session_token,
        username: username,
        password: password,
        licenseKey: license_key,
        hwid: @hwid,
        pcName: pc_name,
      })
      @user = parse_user(data)
      data
    end

    def register(username, password, email: "", display_name: "", license_key: "", pc_name: "")
      require_init!
      fetch_nonce
      data = post_with_nonce("/api/client/register", {
        sessionToken: @session_token,
        username: username,
        password: password,
        email: email,
        displayName: display_name,
        licenseKey: license_key,
        hwid: @hwid,
        pcName: pc_name,
      })
      @user = parse_user(data)
      data
    end

    def license_login(license_key, pc_name: "")
      require_init!
      fetch_nonce
      data = post_with_nonce("/api/client/license", {
        sessionToken: @session_token,
        licenseKey: license_key,
        hwid: @hwid,
        pcName: pc_name,
      })
      @user = parse_user(data)
      data
    end

    # ── Token Validation ────────────────────────────────────────────

    def validate_token(token)
      require_init!
      post("/api/client/token", { sessionToken: @session_token, token: token })
    end

    # ── License Activation ──────────────────────────────────────────

    def activate(license_key)
      require_init!
      fetch_nonce
      post_with_nonce("/api/client/activate", { sessionToken: @session_token, licenseKey: license_key })
    end

    # ── Variables ───────────────────────────────────────────────────

    def get_variable(key)
      require_init!
      data = post("/api/client/variable", { sessionToken: @session_token, key: key })
      Variable.new(key: data["key"], value: data["value"], type: data["type"] || "STRING")
    end

    def get_all_variables
      require_init!
      data = post("/api/client/variables", { sessionToken: @session_token })
      return [] unless data.is_a?(Array)
      data.map { |v| Variable.new(key: v["key"], value: v["value"], type: v["type"] || "STRING") }
    end

    # ── User Variables ──────────────────────────────────────────────

    def get_user_variable(key)
      require_init!
      data = post("/api/client/user-variable", { sessionToken: @session_token, key: key })
      UserVariable.new(key: data["key"], value: data["value"])
    end

    def get_all_user_variables
      require_init!
      data = post("/api/client/user-variables", { sessionToken: @session_token })
      return [] unless data.is_a?(Array)
      data.map { |v| UserVariable.new(key: v["key"], value: v["value"]) }
    end

    def set_user_variable(key, value)
      require_init!
      data = post("/api/client/set-user-variable", { sessionToken: @session_token, key: key, value: value })
      UserVariable.new(key: data["key"], value: data["value"])
    end

    def delete_user_variable(key)
      require_init!
      post("/api/client/delete-user-variable", { sessionToken: @session_token, key: key })
    end

    # ── Files ───────────────────────────────────────────────────────

    def download_file(name)
      require_init!
      uri = URI("#{@base_url}/api/client/file")
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"
      req = Net::HTTP::Post.new(uri.path, { "Content-Type" => "application/json" })
      req.body = JSON.generate({ sessionToken: @session_token, name: name })
      resp = http.request(req)
      data = resp.body.b

      if resp["x-file-encrypted"] == "1" && resp["x-file-nonce"]
        data = decrypt_file_bytes(data, @secret, @session_token, resp["x-file-nonce"])
      end

      data
    end

    def check_update(current_version, file_name: "")
      require_init!
      post("/api/client/check-update", {
        sessionToken: @session_token,
        currentVersion: current_version,
        fileName: file_name,
      })
    end

    # ── Session Management ──────────────────────────────────────────

    def heartbeat
      require_init!
      fetch_nonce
      post_with_nonce("/api/client/heartbeat", { sessionToken: @session_token })
    end

    def check_session
      require_init!
      post("/api/client/check", { sessionToken: @session_token })
    end

    def end_session
      require_init!
      post("/api/client/end", { sessionToken: @session_token })
      @session_token = nil
      @user = nil
    end

    # ── User Info ───────────────────────────────────────────────────

    def get_user
      require_init!
      post("/api/client/user", { sessionToken: @session_token })
    end

    def change_password(current_password, new_password)
      require_init!
      post("/api/client/change-password", {
        sessionToken: @session_token,
        currentPassword: current_password,
        newPassword: new_password,
      })
    end

    def request_reset
      require_init!
      post("/api/client/request-reset", { sessionToken: @session_token })
    end

    # ── Client Log ──────────────────────────────────────────────────

    def log(message, level: "INFO")
      require_init!
      post("/api/client/log", { sessionToken: @session_token, message: message, level: level })
    end

    private

    def require_init!
      raise Error.new("NOT_INITIALIZED", "Call init() before using other methods.") unless @session_token
    end

    def fetch_nonce
      data = post("/api/client/nonce", { sessionToken: @session_token })
      @nonce = data["nonce"]
    end

    def post(path, payload)
      request(path, payload, {})
    end

    def post_with_nonce(path, payload)
      request(path, payload, { "X-Nonce" => @nonce })
    end

    def request(path, payload, extra_headers)
      uri = URI("#{@base_url}#{path}")
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"

      req = Net::HTTP::Post.new(uri.path, {
        "Content-Type" => "application/json",
        "Accept" => "application/json",
      }.merge(extra_headers))
      req.body = JSON.generate(payload)

      resp = http.request(req)
      body = JSON.parse(resp.body)

      unless body["success"]
        err = body["error"] || {}
        raise Error.new(err["code"] || "UNKNOWN", err["message"] || "Request failed")
      end

      body["data"] || body["message"] || {}
    end

    def parse_user(data)
      UserData.new(
        key: data["key"] || "",
        username: data["username"] || "",
        email: data["email"] || "",
        level: data["level"] || 0,
        expires_at: data["expiresAt"],
        metadata: data["metadata"],
      )
    end

    # ── Multi-Layer File Decryption ─────────────────────────────────

    def decrypt_file_bytes(data, app_secret, session_token, nonce_hex)
      nonce = [nonce_hex].pack("H*")
      aes_key, shuffle_key, xor_key = derive_keys(app_secret, session_token, nonce)

      # Layer 4: Rolling XOR
      buf = data.bytes.to_a
      buf.each_index { |i| buf[i] ^= xor_key.bytes[i % xor_key.bytesize] }

      # Layer 3: Fisher-Yates unshuffle
      n = buf.length
      rng = SeededRNG.new(shuffle_key)
      swaps = []
      (n - 1).downto(1) { |i| swaps << (rng.next_val % (i + 1)) }
      (swaps.length - 1).downto(0) do |k|
        i = n - 1 - k
        j = swaps[k]
        buf[i], buf[j] = buf[j], buf[i]
      end

      # Layer 2: AES-256-GCM
      gcm_nonce = buf[0, 12].pack("C*")
      ciphertext_and_tag = buf[12..].pack("C*")
      tag = ciphertext_and_tag[-16..]
      ciphertext = ciphertext_and_tag[0...-16]
      cipher = OpenSSL::Cipher.new("aes-256-gcm")
      cipher.decrypt
      cipher.key = aes_key
      cipher.iv = gcm_nonce
      cipher.auth_tag = tag
      plaintext = cipher.update(ciphertext) + cipher.final

      # Layer 1: zlib decompress
      Zlib::Inflate.inflate(plaintext)
    end

    def derive_keys(app_secret, session_token, nonce)
      master = OpenSSL::HMAC.digest("SHA256", app_secret, "#{session_token}|#{nonce.unpack1("H*")}")
      [0x01, 0x02, 0x03].map { |layer| Digest::SHA256.digest(master + [layer].pack("C")) }
    end

    class SeededRNG
      def initialize(key)
        @state = 0
        if key.bytesize >= 8
          @state = key[0, 8].unpack1("Q<")
        end
        (8...key.bytesize).each do |i|
          @state ^= (key.getbyte(i) & 0xFF) << ((i % 8) * 8)
          @state = (@state * 6364136223846793005 + 1442695040888963407) & 0xFFFFFFFFFFFFFFFF
        end
        @state = 0xDEADBEEFCAFE1234 if @state == 0
      end

      def next_val
        @state = (@state * 6364136223846793005 + 1442695040888963407) & 0xFFFFFFFFFFFFFFFF
        @state
      end
    end
  end
end

File: go/swiftauth.go
// Package swiftauth provides a Go client for the SwiftAuth authentication API.
package swiftauth

import (
	"bytes"
	"compress/zlib"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"os/user"
	"strings"
	"sync"

	"github.com/gorilla/websocket"
)

// SwiftAuthError represents an API error.
type SwiftAuthError struct {
	Code    string
	Message string
}

func (e *SwiftAuthError) Error() string {
	return fmt.Sprintf("[%s] %s", e.Code, e.Message)
}

// AppInfo holds application metadata returned from init.
type AppInfo struct {
	Name       string `json:"appName"`
	Version    string `json:"appVersion"`
	AntiDebug  bool   `json:"antiDebug"`
	AntiVM     bool   `json:"antiVM"`
	LockHwid   bool   `json:"lockHwid"`
	LockIp     bool   `json:"lockIp"`
	LockPcName bool   `json:"lockPcName"`
}

// UserData holds authenticated user info.
type UserData struct {
	Key       string `json:"key"`
	Username  string `json:"username"`
	Email     string `json:"email"`
	Level     int    `json:"level"`
	ExpiresAt string `json:"expiresAt"`
	Metadata  any    `json:"metadata"`
}

// Variable represents an app variable.
type Variable struct {
	Key   string `json:"key"`
	Value string `json:"value"`
	Type  string `json:"type"`
}

// UserVariable represents a user-scoped variable.
type UserVariable struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

// Client is the main SwiftAuth SDK client.
type Client struct {
	baseURL      string
	secret       string
	version      string
	hwid         string
	sessionToken string
	nonce        string
	httpClient   *http.Client

	App  *AppInfo
	User *UserData

	ws          *websocket.Conn
	wsMu        sync.Mutex
	wsCallbacks map[string][]func(map[string]any)
	wsDone      chan struct{}
}

// NewClient creates a new SwiftAuth client.
func NewClient(baseURL, appSecret, appVersion, hwid string) *Client {
	if hwid == "" {
		hostname, _ := os.Hostname()
		u, _ := user.Current()
		uname := "unknown"
		if u != nil {
			uname = u.Username
		}
		hwid = hostname + "-" + uname
	}
	return &Client{
		baseURL:     strings.TrimRight(baseURL, "/"),
		secret:      appSecret,
		version:     appVersion,
		hwid:        hwid,
		httpClient:  &http.Client{},
		wsCallbacks: make(map[string][]func(map[string]any)),
	}
}

// Init initializes a session with the SwiftAuth server.
func (c *Client) Init() (map[string]any, error) {
	data, err := c.post("/api/client/init", map[string]any{
		"secret":  c.secret,
		"version": c.version,
		"hwid":    c.hwid,
	})
	if err != nil {
		return nil, err
	}
	c.sessionToken = str(data, "sessionToken")
	c.App = &AppInfo{
		Name:       str(data, "appName"),
		Version:    str(data, "appVersion"),
		AntiDebug:  boolVal(data, "antiDebug"),
		AntiVM:     boolVal(data, "antiVM"),
		LockHwid:   boolVal(data, "lockHwid"),
		LockIp:     boolVal(data, "lockIp"),
		LockPcName: boolVal(data, "lockPcName"),
	}
	return data, nil
}

func (c *Client) fetchNonce() error {
	data, err := c.post("/api/client/nonce", map[string]any{
		"sessionToken": c.sessionToken,
	})
	if err != nil {
		return err
	}
	c.nonce = str(data, "nonce")
	return nil
}

// Login authenticates with username and password.
func (c *Client) Login(username, password, licenseKey, pcName string) (map[string]any, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	if err := c.fetchNonce(); err != nil {
		return nil, err
	}
	data, err := c.postWithNonce("/api/client/login", map[string]any{
		"sessionToken": c.sessionToken,
		"username":     username,
		"password":     password,
		"licenseKey":   licenseKey,
		"hwid":         c.hwid,
		"pcName":       pcName,
	})
	if err != nil {
		return nil, err
	}
	c.User = parseUser(data)
	return data, nil
}

// Register creates a new user account.
func (c *Client) Register(username, password, email, displayName, licenseKey, pcName string) (map[string]any, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	if err := c.fetchNonce(); err != nil {
		return nil, err
	}
	data, err := c.postWithNonce("/api/client/register", map[string]any{
		"sessionToken": c.sessionToken,
		"username":     username,
		"password":     password,
		"email":        email,
		"displayName":  displayName,
		"licenseKey":   licenseKey,
		"hwid":         c.hwid,
		"pcName":       pcName,
	})
	if err != nil {
		return nil, err
	}
	c.User = parseUser(data)
	return data, nil
}

// LicenseLogin authenticates with a license key only.
func (c *Client) LicenseLogin(licenseKey, pcName string) (map[string]any, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	if err := c.fetchNonce(); err != nil {
		return nil, err
	}
	data, err := c.postWithNonce("/api/client/license", map[string]any{
		"sessionToken": c.sessionToken,
		"licenseKey":   licenseKey,
		"hwid":         c.hwid,
		"pcName":       pcName,
	})
	if err != nil {
		return nil, err
	}
	c.User = parseUser(data)
	return data, nil
}

// ValidateToken validates a pre-auth access token.
func (c *Client) ValidateToken(token string) (map[string]any, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	return c.post("/api/client/token", map[string]any{
		"sessionToken": c.sessionToken,
		"token":        token,
	})
}

// Activate activates a license on the current user.
func (c *Client) Activate(licenseKey string) (map[string]any, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	if err := c.fetchNonce(); err != nil {
		return nil, err
	}
	return c.postWithNonce("/api/client/activate", map[string]any{
		"sessionToken": c.sessionToken,
		"licenseKey":   licenseKey,
	})
}

// GetVariable fetches a single app variable.
func (c *Client) GetVariable(key string) (*Variable, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	data, err := c.post("/api/client/variable", map[string]any{
		"sessionToken": c.sessionToken,
		"key":          key,
	})
	if err != nil {
		return nil, err
	}
	return &Variable{Key: str(data, "key"), Value: str(data, "value"), Type: str(data, "type")}, nil
}

// GetAllVariables fetches all app variables.
func (c *Client) GetAllVariables() ([]Variable, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	data, err := c.post("/api/client/variables", map[string]any{
		"sessionToken": c.sessionToken,
	})
	if err != nil {
		return nil, err
	}
	return parseVariables(data), nil
}

// GetLicenseVariable fetches a single license-scoped variable.
func (c *Client) GetLicenseVariable(key string) (*Variable, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	data, err := c.post("/api/client/license-variable", map[string]any{
		"sessionToken": c.sessionToken,
		"key":          key,
	})
	if err != nil {
		return nil, err
	}
	return &Variable{Key: str(data, "key"), Value: str(data, "value"), Type: str(data, "type")}, nil
}

// GetAllLicenseVariables fetches all license-scoped variables.
func (c *Client) GetAllLicenseVariables() ([]Variable, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	data, err := c.post("/api/client/license-variables", map[string]any{
		"sessionToken": c.sessionToken,
	})
	if err != nil {
		return nil, err
	}
	return parseVariables(data), nil
}

// GetUserVariable fetches a user-scoped variable.
func (c *Client) GetUserVariable(key string) (*UserVariable, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	data, err := c.post("/api/client/user-variable", map[string]any{
		"sessionToken": c.sessionToken,
		"key":          key,
	})
	if err != nil {
		return nil, err
	}
	return &UserVariable{Key: str(data, "key"), Value: str(data, "value")}, nil
}

// GetAllUserVariables fetches all user-scoped variables.
func (c *Client) GetAllUserVariables() ([]UserVariable, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	data, err := c.post("/api/client/user-variables", map[string]any{
		"sessionToken": c.sessionToken,
	})
	if err != nil {
		return nil, err
	}
	return parseUserVariables(data), nil
}

// SetUserVariable sets or updates a user-scoped variable.
func (c *Client) SetUserVariable(key, value string) (*UserVariable, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	data, err := c.post("/api/client/set-user-variable", map[string]any{
		"sessionToken": c.sessionToken,
		"key":          key,
		"value":        value,
	})
	if err != nil {
		return nil, err
	}
	return &UserVariable{Key: str(data, "key"), Value: str(data, "value")}, nil
}

// DeleteUserVariable deletes a user-scoped variable.
func (c *Client) DeleteUserVariable(key string) error {
	if err := c.requireInit(); err != nil {
		return err
	}
	_, err := c.post("/api/client/delete-user-variable", map[string]any{
		"sessionToken": c.sessionToken,
		"key":          key,
	})
	return err
}

// DownloadFile downloads and decrypts a file by name.
func (c *Client) DownloadFile(name string) ([]byte, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	body, _ := json.Marshal(map[string]any{
		"sessionToken": c.sessionToken,
		"name":         name,
	})
	resp, err := c.httpClient.Post(c.baseURL+"/api/client/file", "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.Header.Get("X-File-Encrypted") == "1" {
		nonceHex := resp.Header.Get("X-File-Nonce")
		if nonceHex != "" {
			return decryptFileBytes(raw, c.secret, c.sessionToken, nonceHex)
		}
	}
	return raw, nil
}

// CheckUpdate checks for app version updates.
func (c *Client) CheckUpdate(currentVersion, fileName string) (map[string]any, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	return c.post("/api/client/check-update", map[string]any{
		"sessionToken":   c.sessionToken,
		"currentVersion": currentVersion,
		"fileName":       fileName,
	})
}

// Heartbeat sends a keep-alive heartbeat.
func (c *Client) Heartbeat() (map[string]any, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	if err := c.fetchNonce(); err != nil {
		return nil, err
	}
	return c.postWithNonce("/api/client/heartbeat", map[string]any{
		"sessionToken": c.sessionToken,
	})
}

// CheckSession verifies the session is still valid.
func (c *Client) CheckSession() (map[string]any, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	return c.post("/api/client/check", map[string]any{
		"sessionToken": c.sessionToken,
	})
}

// EndSession ends and invalidates the current session.
func (c *Client) EndSession() error {
	if err := c.requireInit(); err != nil {
		return err
	}
	_, err := c.post("/api/client/end", map[string]any{
		"sessionToken": c.sessionToken,
	})
	c.sessionToken = ""
	c.User = nil
	return err
}

// GetUser gets current user profile info.
func (c *Client) GetUser() (map[string]any, error) {
	if err := c.requireInit(); err != nil {
		return nil, err
	}
	return c.post("/api/client/user", map[string]any{
		"sessionToken": c.sessionToken,
	})
}

// ChangePassword changes the current user's password.
func (c *Client) ChangePassword(currentPassword, newPassword string) error {
	if err := c.requireInit(); err != nil {
		return err
	}
	_, err := c.post("/api/client/change-password", map[string]any{
		"sessionToken":    c.sessionToken,
		"currentPassword": currentPassword,
		"newPassword":     newPassword,
	})
	return err
}

// RequestReset requests a HWID/IP reset.
func (c *Client) RequestReset() error {
	if err := c.requireInit(); err != nil {
		return err
	}
	_, err := c.post("/api/client/request-reset", map[string]any{
		"sessionToken": c.sessionToken,
	})
	return err
}

// Log sends a client-side log entry.
func (c *Client) Log(message, level string) error {
	if err := c.requireInit(); err != nil {
		return err
	}
	if level == "" {
		level = "INFO"
	}
	_, err := c.post("/api/client/log", map[string]any{
		"sessionToken": c.sessionToken,
		"message":      message,
		"level":        level,
	})
	return err
}

// SessionToken returns the current session token.
func (c *Client) SessionToken() string {
	return c.sessionToken
}

// IsInitialized returns true if the client has been initialized.
func (c *Client) IsInitialized() bool {
	return c.sessionToken != ""
}

// ── WebSocket ───────────────────────────────────────────────────────

// On registers a callback for a WebSocket event type.
func (c *Client) On(event string, cb func(map[string]any)) {
	c.wsCallbacks[event] = append(c.wsCallbacks[event], cb)
}

// ConnectWs opens a WebSocket connection.
func (c *Client) ConnectWs() error {
	if err := c.requireInit(); err != nil {
		return err
	}
	wsURL := strings.Replace(c.baseURL, "https://", "wss://", 1)
	wsURL = strings.Replace(wsURL, "http://", "ws://", 1)
	wsURL += "/api/client/ws?token=" + url.QueryEscape(c.sessionToken)

	conn, _, err := websocket.DefaultDialer.Dial(wsURL, nil)
	if err != nil {
		return err
	}
	c.ws = conn
	c.wsDone = make(chan struct{})

	go func() {
		defer close(c.wsDone)
		for {
			_, msg, err := c.ws.ReadMessage()
			if err != nil {
				for _, cb := range c.wsCallbacks["close"] {
					cb(map[string]any{"error": err.Error()})
				}
				return
			}
			var evt map[string]any
			if json.Unmarshal(msg, &evt) == nil {
				eventType, _ := evt["type"].(string)
				for _, cb := range c.wsCallbacks[eventType] {
					cb(evt)
				}
				for _, cb := range c.wsCallbacks["*"] {
					cb(evt)
				}
			}
		}
	}()
	return nil
}

// WsSend sends a JSON message over the WebSocket.
func (c *Client) WsSend(data map[string]any) error {
	c.wsMu.Lock()
	defer c.wsMu.Unlock()
	if c.ws == nil {
		return fmt.Errorf("websocket not connected")
	}
	return c.ws.WriteJSON(data)
}

// WsPing sends a ping message.
func (c *Client) WsPing() error {
	return c.WsSend(map[string]any{"type": "ping"})
}

// WsSetStatus sets the user's online status.
func (c *Client) WsSetStatus(status string) error {
	return c.WsSend(map[string]any{"type": "set_status", "status": status})
}

// WsSendChat sends a chat message.
func (c *Client) WsSendChat(message string) error {
	return c.WsSend(map[string]any{"type": "chat", "message": message})
}

// WsSetTyping sends a typing indicator.
func (c *Client) WsSetTyping(typing bool) error {
	return c.WsSend(map[string]any{"type": "typing", "typing": typing})
}

// WsSetMetadata sets session metadata.
func (c *Client) WsSetMetadata(metadata map[string]any) error {
	return c.WsSend(map[string]any{"type": "set_metadata", "metadata": metadata})
}

// DisconnectWs closes the WebSocket connection.
func (c *Client) DisconnectWs() {
	if c.ws != nil {
		c.ws.Close()
		c.ws = nil
	}
}

// ── Internal ────────────────────────────────────────────────────────

func (c *Client) requireInit() error {
	if c.sessionToken == "" {
		return &SwiftAuthError{Code: "NOT_INITIALIZED", Message: "Call Init() before using other methods."}
	}
	return nil
}

func (c *Client) post(path string, payload map[string]any) (map[string]any, error) {
	return c.request(path, payload, nil)
}

func (c *Client) postWithNonce(path string, payload map[string]any) (map[string]any, error) {
	return c.request(path, payload, map[string]string{"X-Nonce": c.nonce})
}

func (c *Client) request(path string, payload map[string]any, headers map[string]string) (map[string]any, error) {
	body, _ := json.Marshal(payload)
	req, err := http.NewRequest("POST", c.baseURL+path, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result map[string]any
	if err := json.Unmarshal(raw, &result); err != nil {
		return nil, &SwiftAuthError{Code: "PARSE_ERROR", Message: "Invalid server response"}
	}

	success, _ := result["success"].(bool)
	if !success {
		errMap, _ := result["error"].(map[string]any)
		code := "UNKNOWN"
		msg := "Request failed"
		if errMap != nil {
			if c, ok := errMap["code"].(string); ok {
				code = c
			}
			if m, ok := errMap["message"].(string); ok {
				msg = m
			}
		}
		return nil, &SwiftAuthError{Code: code, Message: msg}
	}

	if data, ok := result["data"].(map[string]any); ok {
		return data, nil
	}
	if data, ok := result["data"]; ok {
		// data could be a slice for list endpoints
		return map[string]any{"_list": data}, nil
	}
	if msg, ok := result["message"]; ok {
		return map[string]any{"message": msg}, nil
	}
	return map[string]any{}, nil
}

// ── Helpers ─────────────────────────────────────────────────────────

func str(m map[string]any, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}

func boolVal(m map[string]any, key string) bool {
	if v, ok := m[key].(bool); ok {
		return v
	}
	return false
}

func parseUser(data map[string]any) *UserData {
	u := &UserData{
		Key:       str(data, "key"),
		Username:  str(data, "username"),
		Email:     str(data, "email"),
		ExpiresAt: str(data, "expiresAt"),
		Metadata:  data["metadata"],
	}
	if v, ok := data["level"].(float64); ok {
		u.Level = int(v)
	}
	return u
}

func parseVariables(data map[string]any) []Variable {
	list, ok := data["_list"].([]any)
	if !ok {
		return nil
	}
	var vars []Variable
	for _, item := range list {
		if m, ok := item.(map[string]any); ok {
			vars = append(vars, Variable{Key: str(m, "key"), Value: str(m, "value"), Type: str(m, "type")})
		}
	}
	return vars
}

func parseUserVariables(data map[string]any) []UserVariable {
	list, ok := data["_list"].([]any)
	if !ok {
		return nil
	}
	var vars []UserVariable
	for _, item := range list {
		if m, ok := item.(map[string]any); ok {
			vars = append(vars, UserVariable{Key: str(m, "key"), Value: str(m, "value")})
		}
	}
	return vars
}

// ── Multi-Layer File Decryption ─────────────────────────────────────

type seededRNG struct {
	state uint64
}

func newSeededRNG(key []byte) *seededRNG {
	var state uint64
	if len(key) >= 8 {
		state = binary.LittleEndian.Uint64(key[:8])
	}
	for i := 8; i < len(key); i++ {
		state ^= uint64(key[i]) << ((uint(i) % 8) * 8)
		state = state*6364136223846793005 + 1442695040888963407
	}
	if state == 0 {
		state = 0xDEADBEEFCAFE1234
	}
	return &seededRNG{state: state}
}

func (r *seededRNG) next() uint64 {
	r.state = r.state*6364136223846793005 + 1442695040888963407
	return r.state
}

func deriveKeys(appSecret, sessionToken string, nonce []byte) (aesKey, shuffleKey, xorKey []byte) {
	mac := hmac.New(sha256.New, []byte(appSecret))
	mac.Write([]byte(sessionToken + "|" + hex.EncodeToString(nonce)))
	master := mac.Sum(nil)

	sub := func(layer byte) []byte {
		h := sha256.Sum256(append(master, layer))
		return h[:]
	}

	return sub(0x01), sub(0x02), sub(0x03)
}

func decryptFileBytes(data []byte, appSecret, sessionToken, nonceHex string) ([]byte, error) {
	nonce, err := hex.DecodeString(nonceHex)
	if err != nil {
		return nil, err
	}
	aesKey, shuffleKey, xorKey := deriveKeys(appSecret, sessionToken, nonce)

	// Layer 4: Rolling XOR
	buf := make([]byte, len(data))
	copy(buf, data)
	for i := range buf {
		buf[i] ^= xorKey[i%len(xorKey)]
	}

	// Layer 3: Fisher-Yates unshuffle
	n := len(buf)
	rng := newSeededRNG(shuffleKey)
	swaps := make([]int, 0, n)
	for i := n - 1; i > 0; i-- {
		swaps = append(swaps, int(rng.next()%uint64(i+1)))
	}
	for k := len(swaps) - 1; k >= 0; k-- {
		i := n - 1 - k
		j := swaps[k]
		buf[i], buf[j] = buf[j], buf[i]
	}

	// Layer 2: AES-256-GCM decrypt
	gcmNonce := buf[:12]
	ciphertextAndTag := buf[12:]
	block, err := aes.NewCipher(aesKey)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	plaintext, err := gcm.Open(nil, gcmNonce, ciphertextAndTag, nil)
	if err != nil {
		return nil, err
	}

	// Layer 1: zlib decompress
	r, err := zlib.NewReader(bytes.NewReader(plaintext))
	if err != nil {
		return nil, err
	}
	defer r.Close()
	return io.ReadAll(r)
}
